### PR TITLE
JM - enable photo sharing by email (through share page)

### PR DESCRIPTION
**Pull request by jmangel**

### CODE REVIEW

File: config/routes.rb [hunk 1]
```diff
@@ -4,6 +4,8 @@
   root to: 'images#index'
 
   get '/images/filter/:tag', to: 'images#index'
+  get '/images/share/:id', to: 'images#share', as: 'share'
+  post '/images/share/:id', to: 'images#send_email'
```
Comment (lsq921224) [inline config/routes.rb]: this does not seem RESTful to me because`share` and `send_email` are member functions of images resources.

I would do something like 
```
resources :images, only: [:index, :create, :show, :destroy]
    get :share, on: :member
    post :share, to: 'images#send_mail', on: :member
end
```
and as a result, we should get the routes as
```
get /images/:id/share
get /images/:id/end_email
```
instead

File: config/environments/development.rb [hunk 2]
```diff
@@ -26,11 +26,13 @@
     config.cache_store = :null_store
   end
 
-  # Don't care if the mailer can't send.
-  config.action_mailer.raise_delivery_errors = false
-
   config.action_mailer.perform_caching = false
 
+  config.action_mailer.default_url_options = { host: 'localhost:3000' }
```
Comment (lsq921224) [inline config/environments/development.rb]: These email configurations are recommended to put in its own commits

Comment (lsq921224) [reply config/environments/development.rb]: I just saw your second commit, I would put these changes with the second commit, and swap the order of these two commits

File: config/environments/production.rb [hunk 3]
```diff
@@ -59,7 +59,17 @@
 
   config.action_mailer.default_url_options = { host: ENV['MAILER_HOST'] }
   config.action_mailer.raise_delivery_errors = true
-  config.action_mailer.delivery_method = :sendmail
+  config.action_mailer.delivery_method = :smtp
```
Comment (lsq921224) [inline config/environments/production.rb]: this change should come with the your changes in `development.rb` and `test.rb`, because they all configure the app to be able to send emails

File: app/views/image_mailer/share_email.html.erb [hunk 4]
```diff
@@ -5,8 +5,10 @@
   </head>
   <body>
     <h1>Someone shared an image with you.</h1>
-    <p>
+    <% if @message.present? %>
```
Comment (lsq921224) [inline app/views/image_mailer/share_email.html.erb]: since you already included the optional message in `share_email.text.erb` in your first commit. I would simply squash this commit with that commit. It does not make sense to have two templates, one with message and one without..

File: app/mailers/image_mailer.rb [hunk 5]
```diff
@@ -0,0 +1,10 @@
+class ImageMailer < ApplicationMailer
```
[the remark this answers is not in the document]
Comment (jmangel) [reply app/mailers/image_mailer.rb]: done 👍

File: config/environments/development.rb (hunk 2 above)
Comment (jmangel) [reply config/environments/development.rb]: done 👍

File: app/models/share_email.rb [hunk 6]
```diff
@@ -0,0 +1,3 @@
+class ShareEmail < ApplicationRecord
```
Comment (lsq921224) [inline app/models/share_email.rb]: This works except we don't really need activerecord for this model (It's not necessary to store the email to the db)
Instead, we could do
```
class ShareEmail
  include ActiveModel::Model

  attr_accessor :email_address, :message

  validates_email_format_of :email_address
end
```
